spotlight-client/src/stores/UserStore.ts

```diff
@@ -0,0 +1,116 @@
+// Recidiviz - a data platform for criminal justice reform
+// Copyright (C) 2020 Recidiviz, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+
+import { Auth0Client, Auth0ClientOptions } from "@auth0/auth0-spa-js";
+import { makeAutoObservable, runInAction } from "mobx";
+import qs from "qs";
+import { ERROR_MESSAGES } from "../constants";
+import RootStore from "./RootStore";
+
+type ConstructorProps = {
+  authSettings?: Auth0ClientOptions;
+  isAuthRequired: boolean;
+  rootStore?: RootStore;
+};
+
+/**
+ * Reactive wrapper around Auth0 client.
+ * If auth is disabled for this environment, all users will be immediately authorized.
+ * Otherwise, call `authorize` to retrieve credentials or start login flow.
+ *
+ * @example
+ *
+ * ```js
+ * const store = new UserStore({ isAuthRequired: true, authSettings: { domain, client_id, redirect_uri } });
+ * if (!store.isAuthorized) {
+ *   await store.authorize();
+ *   // this may trigger a redirect to the Auth0 login domain;
+ *   // if we're still here and user has successfully logged in,
+ *   // store.isAuthorized should now be true.
+ * }
+ * ```
+ */
+export default class UserStore {
+  authSettings?: Auth0ClientOptions;
+
+  awaitingVerification: boolean;
+
+  isAuthorized: boolean;
+
+  isLoading: boolean;
+
+  rootStore?: RootStore;
+
+  constructor({ authSettings, isAuthRequired, rootStore }: ConstructorProps) {
+    makeAutoObservable(this);
+
+    this.authSettings = authSettings;
+    this.rootStore = rootStore;
+
+    this.awaitingVerification = false;
+    if (!isAuthRequired) {
+      this.isAuthorized = true;
+      this.isLoading = false;
+    } else {
+      this.isAuthorized = false;
+      this.isLoading = true;
+    }
+  }
+
+  /**
+   * If user already has a valid Auth0 credential, this method will retrieve it
+   * and update class properties accordingly. If not, user will be redirected
+   * to the Auth0 login domain for fresh authentication.
+   * Returns an Error if Auth0 configuration is not present.
+   */
+  async authorize(): Promise<Error | void> {
```

Comment:
Did you happen to see the [Auth0 post](https://auth0.com/blog/building-and-securing-mobx-apps-with-auth0/) about Auth0 x React x MobX? If not, how does your approach here compare to what they described therein? In particular, I want to make sure we are writing as little auth-related logic as possible ourselves. I want to delegate as much as possible either into the auth layer itself or into vetted code written by auth professionals.

Reply:
I did use it as a reference to some degree, but I am going to read it again through this particular lens and report back

Reply:
OK so after another review there are a few things I would change:
- I think I should just call the `createAuth0Client` function instead of initializing it manually, so that it does the Officially Recommended Things (which I still find to be largely opaque and mysterious). whatever reason I had for not doing that originally does not seem to be applicable anymore
- I was a bit hesitant to create a dedicated callback url like they do in this example because I thought the application should be able to authorize you "in place". looking at it again this morning though, I am not so convinced that makes sense, since it requires us to expect that auth0 is going to add certain search params to the url. (I adapted that logic from the COVID site but inspecting that seems unnecessarily brittle in this context.) There should be other ways to make sure we return you to the original page after authorizing you, if in fact we even care to do that

The only other significant difference between this code and that example is that they have you click a button to log in whereas we just trigger it automatically. There is no difference there in the underlying interactions with the Auth0 SDK, it's just a user experience difference.

Thanks for calling this out! Definitely something that got a bit lost in the weeds

Reply:
ok well this turned out to be an unexpected journey .... I did adopt the `createAuth0Client` factory function, which is a definite improvement; in the absence of that the Auth0 SDK was doing some extra redirects to authenticate you which was making for a very choppy experience. 

Creating a callback URL will require us to figure out what the URL routing story will be, which turned out be not very straightforward; there doesn't seem to be any open-source solution out there that Just Works in the way I was looking for (I evaluated like three or four, all of which eventually fell over in Extremely Not Fun ways). I don't think there is anything really wrong with the existing solution for callback handling, since it is also an Auth0-recommended approach and one we have used elsewhere. This could be something to revisit later once we have URL routing set up, but since this only affects staging for now anyway I don't think it's worth spending any more time on right now or adding to the scope of this feature.